src/main/kotlin/org/rust/lang/core/parser/RustParserUtil.kt

```diff
@@ -434,15 +434,28 @@ object RustParserUtil : GeneratedParserUtilBase() {
         return true
     }
 
-    // foo ! ();
-    // ^ this name
-    private fun lookupSimpleMacroName(b: PsiBuilder): String? {
-        return if (b.tokenType == IDENTIFIER) {
-            val nextTokenIsExcl = b.probe {
-                b.advanceLexer()
-                b.tokenType == EXCL
+    // qualifier::foo ! ();
+    //            ^ this name
+    @Suppress("SameParameterValue")
+    private fun getMacroName(b: PsiBuilder, nameTokenIndex: Int): String? {
+        require(nameTokenIndex < 0) {
+            "`getMacroName` assumes that path with macro name is already parsed and the name token is behind of current position"
+        }
+
+        var steps = 0
+        var meaningfulSteps = 0
+
+        while (meaningfulSteps > nameTokenIndex) {
+            steps--
+            val elementType = b.rawLookup(steps) ?: return null
+
+            if (!isWhitespaceOrComment(b, elementType)) {
+                meaningfulSteps--
             }
-            if (nextTokenIsExcl) b.tokenText else null
+        }
```

Comment:
Let's extract it to extension `lookBehind` (similar to existing `lookAhead`) in `org/rust/lang/core/parser/util.kt`. This method would look more clear.

Reply:
Note, `lookAhead` method returns `IElementType` but in this code we need to know raw index as well to get token text. So we can't create something like `fun PsiBuilder.lookBehind(steps: Int): IElementType?` only `fun PsiBuilder.lookBehind(steps: Int): Pair<Int, IElementType>?`. Do you think it's worth it? I don't like such inconsistency

Reply:
Indeed...